formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt

```diff
@@ -0,0 +1,330 @@
+    override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
+
+        if ((encodeByteArrayAsByteString || cbor.alwaysUseByteString)
+            && serializer.descriptor == ByteArraySerializer().descriptor
+        ) {
+            getDestination().encodeByteString(value as ByteArray)
+        } else {
+            encodeByteArrayAsByteString = encodeByteArrayAsByteString || serializer.descriptor.isInlineByteString()
+            super.encodeSerializableValue(serializer, value)
+        }
+    }
+
+    override fun shouldEncodeElementDefault(descriptor: SerialDescriptor, index: Int): Boolean = cbor.encodeDefaults
+
+    protected abstract fun incrementChildren()
+
+    override fun encodeString(value: String) {
+        getDestination().encodeString(value)
+    }
+
+
+    override fun encodeFloat(value: Float) {
+        getDestination().encodeFloat(value)
+    }
+
+
+    override fun encodeDouble(value: Double) {
+        getDestination().encodeDouble(value)
+    }
+
+
+    override fun encodeChar(value: Char) {
+        getDestination().encodeNumber(value.code.toLong())
+    }
+
+
+    override fun encodeByte(value: Byte) {
+        getDestination().encodeNumber(value.toLong())
+    }
+
+
+    override fun encodeShort(value: Short) {
+        getDestination().encodeNumber(value.toLong())
+    }
+
+    override fun encodeInt(value: Int) {
+        getDestination().encodeNumber(value.toLong())
+    }
+
+
+    override fun encodeLong(value: Long) {
+        getDestination().encodeNumber(value)
+    }
+
+
+    override fun encodeBoolean(value: Boolean) {
+        getDestination().encodeBoolean(value)
+    }
+
+
+    override fun encodeNull() {
+        if (isClass) getDestination().encodeEmptyMap()
+        else getDestination().encodeNull()
+    }
+
+    @OptIn(ExperimentalSerializationApi::class) // KT-46731
+    override fun encodeEnum(
+        enumDescriptor: SerialDescriptor,
+        index: Int
+    ) {
+        getDestination().encodeString(enumDescriptor.getElementName(index))
+    }
+
+    override fun encodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
+        val destination = getDestination()
+        isClass = descriptor.getElementDescriptor(index).kind == StructureKind.CLASS
+        encodeByteArrayAsByteString = descriptor.isByteString(index)
+
+        val name = descriptor.getElementName(index)
+
+
+        if (!descriptor.hasArrayTag()) {
+            if (cbor.writeKeyTags) descriptor.getKeyTags(index)?.forEach { destination.encodeTag(it) }
+
+            if ((descriptor.kind !is StructureKind.LIST) && (descriptor.kind !is StructureKind.MAP) && (descriptor.kind !is PolymorphicKind)) {
+                //indices are put into the name field. we don't want to write those, as it would result in double writes
+                if (cbor.preferCborLabelsOverNames) {
+                    descriptor.getCborLabel(index)?.let { destination.encodeNumber(it) }
```

Comment:
It seems that in case we have `preferLabelsOverNames = true` but `@CborLabel` annotation is missing from the property, we won't write anything. Looks like a bug — shouldn't we write name as a fallback?

Reply:
fixed in 5697e3fea29719b8767c17182aab5d5ada5e6276 including a test case